lib/Fmt_ast.ml

```diff
@@ -538,12 +538,38 @@ let fmt_quoted_string key ext s = function
         (str (Format_.sprintf "|%s}" delim))
         (str s)
 
-let fmt_type_var s =
-  str "'"
-  (* [' a'] is a valid type variable, the space is required to not lex as a
-     char. https://github.com/ocaml/ocaml/pull/2034 *)
-  $ fmt_if (String.length s > 1 && Char.equal s.[1] '\'') " "
-  $ str s
+let type_var_has_layout_annot (_, layout_opt) = Option.is_some layout_opt
+
+let layout_to_string = function
+  | Any -> "any"
+  | Value -> "value"
+  | Void -> "void"
+  | Immediate64 -> "immediate64"
+  | Immediate -> "immediate"
+  | Float64 -> "float64"
+
+let fmt_layout_str ~c ~loc string =
+  fmt "@ :@ " $ Cmts.fmt c loc @@ str string
+
+let fmt_layout c l = fmt_layout_str ~c ~loc:l.loc (layout_to_string l.txt)
+
+let fmt_type_var ~have_tick c s =
+  let {txt= name_opt; loc= name_loc}, layout_opt = s in
+  let var_name = Option.value name_opt ~default:"_" in
+  Cmts.fmt c name_loc
+  @@ ( fmt_if_k
+         (Option.is_some name_opt && have_tick)
+         ( str "'"
+         (* [' a'] is a valid type variable, the space is required to not lex
+            as a char. https://github.com/ocaml/ocaml/pull/2034 *)
+         $ fmt_if
+             (String.length var_name > 1 && Char.equal var_name.[1] '\'')
+             " " )
+     $ str var_name )
+  $ Option.value_map layout_opt ~default:noop ~f:(fmt_layout c)
```

Comment:
This is still a little awkward: there is no nice description of the information that is held in `var_name` -- at least not without consulting `have_tick`. May I suggest

```suggestion
let fmt_type_var ~have_tick c s =
  let {txt= name_opt; loc= name_loc}, layout_opt = s in
  Cmts.fmt c name_loc @@
  (match name_opt with
  | None -> str "_"
  | Some var_name ->
    fmt_if_k have_tick
    ( str "'"
         (* [' a'] is a valid type variable, the space is required to not lex
            as a char. https://github.com/ocaml/ocaml/pull/2034 *)
         $ fmt_if
             (String.length var_name > 1 && Char.equal var_name.[1] '\'')
             " " ) $
             str var_name)
  $ Option.value_map layout_opt ~default:noop ~f:(fmt_layout c)
```             

It just seems to be that something like the above is easier to check for correctness. (I did not test this code at all.)